File: src/main/java/ru/arc/stock/ConfigStock.java
package ru.arc.stock;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import com.google.gson.Gson;
import lombok.AllArgsConstructor;
import lombok.Data;
import org.bukkit.Material;
import ru.arc.board.ItemIcon;

@Data
@AllArgsConstructor
public class ConfigStock {

    String symbol;
    String display;
    List<String> lore;
    ItemIcon icon;
    int maxLeverage;
    Stock.Type type;

    public Stock toStock(double price, double dividend, long lastUpdated, long lastTimeDividend){
        return new Stock(
                symbol,
                price,
                dividend,
                lastUpdated,
                display,
                lore,
                icon,
                lastTimeDividend,
                maxLeverage,
                type
        );
    }

    public static ConfigStock deserialize(Map<String, Object> map) {

        Object o = map.get("icon");
        ItemIcon icon = ItemIcon.of(Material.PAPER, 0);
        if(o instanceof String str){
            icon = new Gson().fromJson(str, ItemIcon.class);
        } else if (o instanceof Map<?, ?> iconMap) {
            String uuidString = iconMap.get("headUuid") instanceof String s ? s : null;
            UUID uuid = uuidString == null || uuidString.length() < 36 ? null : UUID.fromString(uuidString);
            String materialStr = iconMap.get("material") instanceof String ms ? ms : "PAPER";
            int data = iconMap.get("data") instanceof Number n ? n.intValue() : 0;
            icon = new ItemIcon(Material.valueOf(materialStr.toUpperCase()), uuid, data);
        }




        Object l = map.getOrDefault("lore", List.of("lore"));
        List<String> lore = new ArrayList<>();
        if (l instanceof String s) {
            lore.add(s);
        } else if (l instanceof Collection<?> list) {
            for (Object item : list) {
                if (item instanceof String str) {
                    lore.add(str);
                }
            }
        }

        return new ConfigStock(
                (String) map.get("symbol"),
                (String) map.getOrDefault("display", "display"),
                lore,
                icon,
                ((Number) map.getOrDefault("maxLeverage", 10000)).intValue(),
                Stock.Type.valueOf(((String) map.getOrDefault("type", "STOCK")).toUpperCase())
        );
    }

}

File: src/main/kotlin/ru/arc/stock/ConfigStock.kt
package ru.arc.stock

import com.google.gson.Gson
import org.bukkit.Material
import ru.arc.board.ItemIcon
import java.util.UUID

data class ConfigStock(
    var symbol: String,
    var display: String,
    var lore: List<String>,
    var icon: ItemIcon,
    var maxLeverage: Int,
    var type: Stock.Type,
) {
    fun toStock(price: Double, dividend: Double, lastUpdated: Long, lastTimeDividend: Long): Stock =
        Stock(
            symbol = symbol,
            price = price,
            dividend = dividend,
            lastUpdated = lastUpdated,
            display = display,
            lore = lore,
            icon = icon,
            lastTimeDividend = lastTimeDividend,
            maxLeverage = maxLeverage,
            type = type,
        )

    companion object {
        private val gson = Gson()

        @JvmStatic
        fun deserialize(map: Map<String, Any>): ConfigStock {
            val icon = parseIcon(map["icon"])
            val lore = parseLore(map.getOrDefault("lore", listOf("lore")))
            return ConfigStock(
                symbol = map["symbol"] as? String ?: "",
                display = map.getOrDefault("display", "display") as? String ?: "display",
                lore = lore,
                icon = icon,
                maxLeverage = (map.getOrDefault("maxLeverage", 10000) as? Number)?.toInt() ?: 10000,
                type = Stock.Type.valueOf(
                    ((map.getOrDefault("type", "STOCK") as? String) ?: "STOCK").uppercase()
                ),
            )
        }

        private fun parseIcon(o: Any?): ItemIcon = when {
            o is String -> gson.fromJson(o, ItemIcon::class.java)
            o is Map<*, *> -> {
                val uuidString = o["headUuid"] as? String
                val uuid = if (uuidString != null && uuidString.length >= 36) UUID.fromString(uuidString) else null
                val materialStr = (o["material"] as? String) ?: "PAPER"
                val data = (o["data"] as? Number)?.toInt() ?: 0
                ItemIcon(Material.valueOf(materialStr.uppercase()), uuid, data)
            }
            else -> ItemIcon.of(Material.PAPER, 0)
        }

        private fun parseLore(l: Any?): List<String> = when (l) {
            is String -> listOf(l)
            is Collection<*> -> l.filterIsInstance<String>()
            else -> listOf("lore")
        }
    }
}
